[MP Rewrite Part 7] Add global obstacles; fix dynamic obstacle properties

kylestach opened the pull request:
This PR adds global and goal-zone obstacles to the planning stack. It also fixes how we deal with dynamic obstacles by essentially treating them as static after they reach the end of their trajectory.

kylestach committed:
Add global obstacles

kylestach committed:
Add goal zone obstacles

kylestach committed:
Make robots avoid ball with correct inflation radius

kylestach committed:
Reduce number of attempts to avoid dynamic obstacles

This improves the worst-case runtime

kylestach committed:
Formatting

JNeiger commented on soccer/gameplay/GameplayModule.cpp:
```diff
@@ -314,6 +314,9 @@ void Gameplay::GameplayModule::run() {
         _context->world_state.ball.position);
 
+    _context->globalObstacles = globalObstacles();
+    _context->goalZoneObstacles = goalZoneObstacles();
```
Does it make sense to have the global obstacles set here? It doesn't really feel like it fits. For the sake of this PR though, this is fine

kylestach replied:
Yeah this has always been where they're created but I think it really belongs with the field geometry calculations.

oswinso commented on soccer/planning/low_level/CreatePath.cpp:
```diff
@@ -52,7 +52,7 @@ Trajectory rrt(const LinearMotionInstant& start,
     ShapeSet obstacles = static_obstacles;
     Trajectory path{{}};
-    constexpr int attemptsToAvoidDynamics = 10;
+    constexpr int attemptsToAvoidDynamics = 3;
```
I guess probably make this a param as well?

oswinso reviewed (approved):
lgtm

kylestach committed:
[MP Rewrite Part 8] Optimize geometry checking (#1534)

* Optimize rectangle intersection fucntion

* Make gameplay geometry proper Rects rather than polygons

* Oops, all typos

[MP Rewrite Part 9] Improve test coverage for motion planning primitives (#1536)

* Write Bezier path tests

* Improve automated testing

* Fix some primitives for tests

[MP Rewrite Part 10] Overall improvements to planners (#1537)

* Add angle overrides and fix pivot planner. Small tweaks to settle.

* Add TODOs for curvature limiting/Bezier spline issues

* Make planner tests deterministic

* Clean up and comment logic in settle planner

* Remove empty CreatePathTest.cpp

* Re-add RRT tests in CreatePathTest

* Make settle planner pick best buffer

This fixes the edge case where the planner can't find an path that finds
the ball quickly enough because the only way to catch it doesn't have
enough buffer time (i.e. immediately before it is caught).

[MP Rewrite Part 11] Miscellaneous formatting and cleanup (#1540)

* Formatting

* More formatting

* Tidy fixes

* Straggling formatting

* Formatting

* Rename low_level to primitives

* Comment on cached_pivot_point

[MP Rewrite Part 12] Port the line kick planner (#1541)

* Add in sketchy line kick

* Formatting

* Tidy

[MP Rewrite Part 13] Stop play fixes and holding trajectory (#1542)

* Fix stopped play and hold in place at end of path move

* Formatting

* Dumb NOLINT

* Style again
